src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WorkerThread.cs

```diff
@@ -0,0 +1,164 @@
+                        // if there is work to do.
+                        if (ThreadPool.GetQueuedWorkItems().Any())
+                        {
+                            MaybeAddWorkingWorker();
+                        }
+                    }
+
+                    ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                    while (true)
+                    {
+                        if (counts.numExistingThreads == counts.numProcessingWork)
+                        {
+                            break;
+                        }
+
+                        ThreadCounts newCounts = counts;
+                        newCounts.numExistingThreads--;
+                        newCounts.numThreadsGoal = Math.Max(s_minThreads, Math.Min(newCounts.numExistingThreads, newCounts.numThreadsGoal));
+                        ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, counts);
+                        if (oldCounts == counts)
+                        {
+                            HillClimbing.ThreadPoolHillClimber.ForceChange(newCounts.numThreadsGoal, HillClimbing.StateOrTransition.ThreadTimedOut);
+                            // TODO: Event:  Worker Thread stop event
+                            return;
+                        }
+                        counts = oldCounts;
+                    } 
+                }
+            }
+
+            internal static void MaybeAddWorkingWorker()
+            {
+                ThreadCounts counts = ThreadCounts.VolatileReadCounts(ref s_aligned.counts);
+                ThreadCounts newCounts;
+                while (true)
+                {
+                    newCounts = counts;
+                    newCounts.numProcessingWork = Math.Max(counts.numProcessingWork, Math.Min((short)(counts.numProcessingWork + 1), counts.numThreadsGoal));
+                    newCounts.numExistingThreads = Math.Max(counts.numExistingThreads, newCounts.numProcessingWork);
+                    
+                    if(newCounts == counts)
+                    {
+                        return;
+                    }
+
+                    ThreadCounts oldCounts = ThreadCounts.CompareExchangeCounts(ref s_aligned.counts, newCounts, counts);
+
+                    if(oldCounts == counts)
+                    {
+                        break;
+                    }
+
+                    counts = oldCounts;
+                }
+
+                int toCreate = newCounts.numExistingThreads - counts.numExistingThreads;
+                int toRelease = newCounts.numProcessingWork - counts.numProcessingWork;
+
+                if(toRelease > 0)
+                {
+                    s_semaphore.Release(toRelease);
+                }
+
+                for (int i = 0; i < toCreate; i++)
+                {
+                    CreateWorkerThread();
+                }
+            }
+
+            internal static bool ShouldStopProcessingWorkNow()
```

Comment:
Would be nice to add a comment describing in what situation this method will determine that a thread should stop working and what causes it